File: aspire-eshop-react/src/services/cartService.ts
import { v4 as uuidv4 } from 'uuid';
import { cartApi } from './api';
import { CartItem } from '../types';

class CartService {
  private sessionId: string;
  private listeners: (() => void)[] = [];

  constructor() {
    // Get or create session ID
    this.sessionId = this.getOrCreateSessionId();
  }

  private getOrCreateSessionId(): string {
    let sessionId = localStorage.getItem('cartSessionId');
    if (!sessionId) {
      sessionId = uuidv4();
      localStorage.setItem('cartSessionId', sessionId);
    }
    return sessionId;
  }

  public getSessionId(): string {
    return this.sessionId;
  }

  public subscribe(listener: () => void): () => void {
    this.listeners.push(listener);
    return () => {
      this.listeners = this.listeners.filter(l => l !== listener);
    };
  }

  private notifyListeners(): void {
    this.listeners.forEach(listener => listener());
  }

  public async getCartItems(): Promise<CartItem[]> {
    try {
      return await cartApi.get(this.sessionId);
    } catch (error) {
      console.error('Error getting cart items:', error);
      return [];
    }
  }

  public async addToCart(productId: number, quantity: number = 1): Promise<CartItem[]> {
    try {
      const result = await cartApi.add(this.sessionId, { productId, quantity });
      this.notifyListeners();
      return result;
    } catch (error) {
      console.error('Error adding to cart:', error);
      throw error;
    }
  }

  public async updateCartItem(itemId: number, quantity: number): Promise<CartItem[]> {
    try {
      const result = await cartApi.update(this.sessionId, itemId, { quantity });
      this.notifyListeners();
      return result;
    } catch (error) {
      console.error('Error updating cart item:', error);
      throw error;
    }
  }

  public async removeFromCart(itemId: number): Promise<void> {
    try {
      await cartApi.remove(this.sessionId, itemId);
      this.notifyListeners();
    } catch (error) {
      console.error('Error removing from cart:', error);
      throw error;
    }
  }

  public async clearCart(): Promise<void> {
    try {
      await cartApi.clear(this.sessionId);
      this.notifyListeners();
    } catch (error) {
      console.error('Error clearing cart:', error);
      throw error;
    }
  }

  public async getCartTotal(): Promise<number> {
    try {
      const items = await this.getCartItems();
      return items.reduce((total, item) => total + (item.product.price * item.quantity), 0);
    } catch (error) {
      console.error('Error calculating cart total:', error);
      return 0;
    }
  }

  public async getCartItemCount(): Promise<number> {
    try {
      const items = await this.getCartItems();
      return items.reduce((count, item) => count + item.quantity, 0);
    } catch (error) {
      console.error('Error getting cart item count:', error);
      return 0;
    }
  }
}

export const cartService = new CartService();

File: aspire-eshop-minimart.Web/Services/CartService.cs
using Microsoft.JSInterop;

namespace aspire_eshop_minimart.Web.Services;

public class CartService
{
    private readonly CartApiClient _cartApiClient;
    private readonly IJSRuntime _jsRuntime;
    private string? _sessionId;
    
    public event Action? OnCartChanged;
    
    public CartService(CartApiClient cartApiClient, IJSRuntime jsRuntime)
    {
        _cartApiClient = cartApiClient;
        _jsRuntime = jsRuntime;
    }

    public async Task<string> GetSessionIdAsync()
    {
        if (string.IsNullOrEmpty(_sessionId))
        {
            try
            {
                // Try to get session ID from localStorage
                _sessionId = await _jsRuntime.InvokeAsync<string>("localStorage.getItem", "cartSessionId");
                
                if (string.IsNullOrEmpty(_sessionId))
                {
                    // Generate new session ID and store it
                    _sessionId = Guid.NewGuid().ToString();
                    await _jsRuntime.InvokeVoidAsync("localStorage.setItem", "cartSessionId", _sessionId);
                }
            }
            catch
            {
                // Fallback if localStorage is not available
                _sessionId = Guid.NewGuid().ToString();
            }
        }
        
        return _sessionId;
    }

    public async Task<CartItem[]> GetCartItemsAsync()
    {
        try
        {
            var sessionId = await GetSessionIdAsync();
            return await _cartApiClient.GetCartAsync(sessionId);
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error getting cart items: {ex.Message}");
            return [];
        }
    }

    public async Task AddToCartAsync(int productId, int quantity = 1)
    {
        try
        {
            var sessionId = await GetSessionIdAsync();
            await _cartApiClient.AddToCartAsync(sessionId, productId, quantity);
            OnCartChanged?.Invoke();
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error adding to cart: {ex.Message}");
            throw; // Re-throw so UI can handle the error
        }
    }

    public async Task UpdateCartItemAsync(int itemId, int quantity)
    {
        try
        {
            var sessionId = await GetSessionIdAsync();
            await _cartApiClient.UpdateCartItemAsync(sessionId, itemId, quantity);
            OnCartChanged?.Invoke();
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error updating cart item: {ex.Message}");
            throw; // Re-throw so UI can handle the error
        }
    }

    public async Task RemoveFromCartAsync(int itemId)
    {
        try
        {
            var sessionId = await GetSessionIdAsync();
            await _cartApiClient.RemoveFromCartAsync(sessionId, itemId);
            OnCartChanged?.Invoke();
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error removing from cart: {ex.Message}");
            throw; // Re-throw so UI can handle the error
        }
    }

    public async Task ClearCartAsync()
    {
        try
        {
            var sessionId = await GetSessionIdAsync();
            await _cartApiClient.ClearCartAsync(sessionId);
            OnCartChanged?.Invoke();
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error clearing cart: {ex.Message}");
            throw; // Re-throw so UI can handle the error
        }
    }

    public async Task<decimal> GetCartTotalAsync()
    {
        try
        {
            var items = await GetCartItemsAsync();
            return items.Sum(item => item.Product.Price * item.Quantity);
        }
        catch
        {
            return 0;
        }
    }

    public async Task<int> GetCartItemCountAsync()
    {
        try
        {
            var items = await GetCartItemsAsync();
            return items.Sum(item => item.Quantity);
        }
        catch
        {
            return 0;
        }
    }
}